flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java

```diff
@@ -0,0 +1,416 @@
+
+		/** The file channel backing the memory mapped file. */
+		private final FileChannel file;
+
+		/** The path of the memory mapped file. */
+		private final Path filePath;
+
+		/** The offset where the next mapped region should start. */
+		private long nextMappingOffset;
+
+		/** The size of each mapped region. */
+		private final long mappingSize;
+
+		MemoryMappedBuffers(Path filePath, FileChannel file, int maxSizePerByteBuffer) throws IOException {
+			this.filePath = filePath;
+			this.file = file;
+			this.mappingSize = alignSize(maxSizePerByteBuffer);
+			this.fullBuffers = new ArrayList<>(4);
+
+			rollOverToNextBuffer();
+		}
+
+		void writeBuffer(Buffer buffer) throws IOException {
+			if (currentBuffer.writeBuffer(buffer)) {
+				return;
+			}
+
+			rollOverToNextBuffer();
+
+			if (!currentBuffer.writeBuffer(buffer)) {
+				throwTooLargeBuffer(buffer);
+			}
+		}
+
+		List<ByteBuffer> getFullBuffers() {
+			return fullBuffers.stream()
+					.map(ByteBuffer::slice)
+					.collect(Collectors.toList());
+		}
+
+		/**
+		 * Finishes the current region and prevents further writes.
+		 */
+		void finishWrite() throws IOException {
+			fullBuffers.add(currentBuffer.complete());
+			currentBuffer = null; // fail further writes fast
+			file.close(); // won't map further regions from now on
+		}
+
+		/**
+		 * Unmaps the file from memory and deletes the file.
+		 * After calling this method, access to any ByteBuffer obtained from this instance
+		 * will cause a segmentation fault.
+		 */
+		void unmapAndDispose() throws IOException {
+			IOUtils.closeQuietly(file); // in case we dispose before finishing writes
+
+			for (ByteBuffer bb : fullBuffers) {
+				PlatformDependent.freeDirectBuffer(bb);
+			}
+			fullBuffers.clear();
+
+			if (currentBuffer != null) {
+				PlatformDependent.freeDirectBuffer(currentBuffer.complete());
+				currentBuffer = null;
+			}
+
+			// To make this compatible with all versions of Windows, we must wait with
+			// deleting the file until it is unmapped.
+			// See also https://stackoverflow.com/questions/11099295/file-flag-delete-on-close-and-memory-mapped-files/51649618#51649618
+
+			Files.delete(filePath);
+		}
+
+		/**
+		 * Gets the number of bytes of all written data (including the metadata in the buffer headers).
+		 */
+		long getSize() {
+			long size = 0L;
+			for (ByteBuffer bb : fullBuffers) {
+				size += bb.capacity();
+			}
+			if (currentBuffer != null) {
+				size += currentBuffer.getNumBytes();
+			}
+			return size;
+		}
+
+		private void rollOverToNextBuffer() throws IOException {
+			if (currentBuffer != null) {
+				// we need to remember the original buffers, not any slices.
+				// slices have no cleaner, which we need to trigger explicit unmapping
+				fullBuffers.add(currentBuffer.complete());
+			}
+
+			final ByteBuffer mapped = file.map(MapMode.READ_WRITE, nextMappingOffset, mappingSize);
+			currentBuffer = new Writer(mapped);
+			nextMappingOffset += mappingSize;
```

Comment:
The previous mapped bytebuffer may not be full, this may lead to file holes between data blocks, though the read view can be created correctly, is it by design.

Reply:
That is by design. If you fill the buffers fully in the writer, the reader need to stitch together buffers across boundaries, which I wanted to avoid. The complexity seems lower this way and the "holes" in the file should be very small, meaning less than one buffer size.